policy/policy_handler/async.go

```diff
@@ -0,0 +1,78 @@
+package policy_handler
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"fmt"
+
+	"github.com/atomist-skills/go-skill"
+	"github.com/atomist-skills/go-skill/policy/data"
+	"olympos.io/encoding/edn"
+)
+
+const eventNameAsyncQuery = data.AsyncQueryName // these must match for the event handler to be registered
+
+// WithAsync will enable async graphql queries for the EventHandler.
+// When used, data.QueryResponse#AsyncRequestMade will be true when performed asynchronously.
+// If that flag is set, the policy evaluator should terminate early with no results.
+// It will be automatically retried once the async query results are returned.
+func WithAsync() Opt {
+	return func(h *EventHandler) {
+		h.subscriptionNames = append(h.subscriptionNames, eventNameAsyncQuery)
+		h.subscriptionDataProviders = append(h.subscriptionDataProviders, getAsyncSubscriptionData)
+		h.dataSourceProviders = append(h.dataSourceProviders, buildAsyncDataSources)
+	}
+}
+
+func getAsyncSubscriptionData(ctx context.Context, req skill.RequestContext) ([][]edn.RawMessage, skill.Configuration, error) {
+	if req.Event.Context.AsyncQueryResult.Name != eventNameAsyncQuery {
+		return nil, skill.Configuration{}, nil
+	}
+
+	metaEdn, err := b64.StdEncoding.DecodeString(req.Event.Context.AsyncQueryResult.Metadata)
+	if err != nil {
+		return nil, skill.Configuration{}, fmt.Errorf("failed to decode async metadata: %w", err)
+	}
+
+	var metadata data.AsyncResultMetadata
+	err = edn.Unmarshal(metaEdn, &metadata)
+	if err != nil {
+		return nil, skill.Configuration{}, fmt.Errorf("failed to unmarshal async metadata: %w", err)
+	}
+
+	return metadata.SubscriptionResults, req.Event.Context.AsyncQueryResult.Configuration, nil
+}
+
+// buildAsyncDataSources always returns at least a data.AsyncDataSource,
+// but also will return a data.FixedDataSource containing the event payload when applicable
```

Comment:
Currently there is no point returning both Async and Fixed - we only support a single Async query.

Reply:
> we only support a single Async query

This is not true anymore, I've expanded the async data source to be capable of storing results in the metadata, allowing it to carry over results from one async query to the next. https://github.com/atomist-skills/go-skill/blob/policy-eval-generification/policy/data/gql_async.go#L75-L80

But the comment is untrue and from an older revision I made :sweat_smile:

Reply:
I'm wary about how well that will work though. The metadata is intended to be small and storing query results in it is likely to start hitting platform limits - it might be simpler to continue to enforce a single async query at least for now instead.

Reply:
I've moved the ability to do multiple async queries behind an explicit opt-in with a comment detailing the metadata length risk in c6c173958d5a8ffcc38f2c7048e1564590e25f13

single async query is now enabled by default